node/core/approval-voting/src/lib.rs

```diff
@@ -539,15 +542,54 @@ struct ApprovalStatus {
 	block_tick: Tick,
 }
 
+enum ApprovalOutcome {
+	Approved,
+	Failed(InvalidCandidate),
+}
+
+struct ApprovalState {
+	candidate_hash: CandidateHash,
+	status: ApprovalOutcome,
+}
+
+impl ApprovalState {
+	fn approved(candidate_hash: CandidateHash) -> Self {
+		Self{ candidate_hash, status: ApprovalOutcome::Approved }
+	}
+	fn failed(candidate_hash: CandidateHash, reason: InvalidCandidate) -> Self {
+		Self { candidate_hash, status: ApprovalOutcome::Failed(reason) }
+	}
+}
+
+struct CheckState {
+	remote_handle: Timeout<RemoteHandle<ApprovalState>>,
+	relay_block_hashes: Vec<Hash>,
+}
+
+impl CheckState {
+	fn new(remote_handle: Timeout<RemoteHandle<ApprovalState>>) -> Self {
+		Self { remote_handle, relay_block_hashes: vec![] }
+	}
+
+	fn insert_relay_block_hash(&mut self, relay_block_hash: Hash) {
+		self.relay_block_hashes.push(relay_block_hash);
+	}
+}
+
 struct State<T> {
 	session_window: RollingSessionWindow,
 	keystore: Arc<LocalKeystore>,
 	slot_duration_millis: u64,
 	db: T,
 	clock: Box<dyn Clock + Send + Sync>,
 	assignment_criteria: Box<dyn AssignmentCriteria + Send + Sync>,
+	approvals_cache: lru::LruCache<CandidateHash, ApprovalOutcome>,
```

Comment:
I recommend:

1. Taking this out of the state which is shared among all pure functions. The higher-level code already handles `Action::LaunchApproval`
2. Factoring all the pieces into a `struct CurrentCheckingSet`

Writing down the API we discussed for `CurrentlyCheckingSet`

```rust
struct CurrentlyCheckingSet {
    awaiting_set: HashMap<CandidateHash, Vec<BlockHash>>,
    remote_work: FuturesUnordered<WrappedRemoteHandlesWithTimeouts>, // each returns candidate hash.
}
impl CurrentlyCheckingSet {
    // Insert a (block, candidate) pair to the set. If the candidate is already known, adds the block to the awaiting set
    // of the candidate without invoking `launch_work`.
    ///
    // If not, this invokes `launch_work` and multiplexes the remote handle along with all others stored and adds
    // block hash to the awaiting set of the candidate.
    fn insert(
        &mut self, 
        candidate_hash: CandidateHash,
        block_hash: BlockHash,
        launch_work: impl FnOnce() -> RemoteHandle<ApprovedOrFailed>,
    );
    async fn await_next_checked(&mut self) -> (CandidateHash, Vec<BlockHash>, ApprovedOrFaile)
}
```